raft.go

```diff
@@ -359,54 +441,91 @@ func (r *Raft) VerifyLeader() Future {
 	}
 }
 
-// AddPeer is used to add a new peer into the cluster. This must be
-// run on the leader or it will fail.
+// TODO: implement GetConfiguration()
+
+// AddPeer (deprecated) is used to add a new peer into the cluster. This must be
+// run on the leader or it will fail. Use AddVoter/AddNonvoter instead.
 func (r *Raft) AddPeer(peer string) Future {
-	logFuture := &logFuture{
-		log: Log{
-			Type: LogAddPeer,
-			peer: peer,
-		},
-	}
-	logFuture.init()
-	select {
-	case r.applyCh <- logFuture:
-		return logFuture
-	case <-r.shutdownCh:
-		return errorFuture{ErrRaftShutdown}
-	}
+	return r.AddVoter(peer, peer, 0, 0)
 }
 
-// RemovePeer is used to remove a peer from the cluster. If the
+// RemovePeer (deprecated) is used to remove a peer from the cluster. If the
 // current leader is being removed, it will cause a new election
 // to occur. This must be run on the leader or it will fail.
+// Use RemoveServer instead.
 func (r *Raft) RemovePeer(peer string) Future {
-	logFuture := &logFuture{
-		log: Log{
-			Type: LogRemovePeer,
-			peer: peer,
-		},
-	}
-	logFuture.init()
-	select {
-	case r.applyCh <- logFuture:
-		return logFuture
-	case <-r.shutdownCh:
-		return errorFuture{ErrRaftShutdown}
-	}
+	return r.RemoveServer(peer, 0, 0)
 }
 
-// SetPeers is used to forcibly replace the set of internal peers and
-// the peerstore with the ones specified. This can be considered unsafe.
-func (r *Raft) SetPeers(p []string) Future {
-	peerFuture := &peerFuture{
-		peers: p,
-	}
-	peerFuture.init()
+// AddVoter will add the given server to the cluster as a staging server. If the
+// server is already in the cluster as a voter, this does nothing. This must be
+// run on the leader or it will fail. The leader will promote the staging server
+// to a voter once that server is ready. If nonzero, prevIndex is the index of
+// the only configuration upon which this change may be applied; if another
+// configuration entry has been added in the meantime, this request will fail.
+// If nonzero, timeout is how long this server should wait before the
+// configuration change log entry is appended.
+func (r *Raft) AddVoter(guid string, address string, prevIndex uint64, timeout time.Duration) IndexFuture {
```

Comment:
Exactly - `type ServerGUID string` would work, or even `type ServerID string`. It's a good idea to do the same for `Address` so we can be super clear which is which, especially as we make the transition.